File: algorithms/19.RemoveNthNodeFromEnd/python/main.py
class ListNode(object):
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
        
class Solution(object):
    def removeNthFromEnd(self, head, n):
        """
        :type head: ListNode
        :type n: int
        :rtype: ListNode
        """
        if head == None: return None
        nodes = []
        cursor = head
        i = 0
        while cursor != None:
            nodes.append(cursor)
            i += 1
            cursor = cursor.next
        nodes.append(None)

        t = i - n
        if t == 0: return head.next
        nodes[t-1].next = nodes[t+1]

        return head

if __name__ == '__main__':
  s = Solution()
  head = ListNode(1, ListNode(2, ListNode(3, ListNode(4, ListNode(5)))))
  head = s.removeNthFromEnd(head, 2)
  while head != None:
    print(head.val)
    head = head.next
                      

File: algorithms/19.RemoveNthNodeFromEnd/c/main.c
#include <stdio.h>
#include <stdlib.h>

struct ListNode {
	int val;
	struct ListNode *next;
};
 
struct ListNode* removeNthFromEnd(struct ListNode* head, int n){
    struct ListNode *nodes[31];
    struct ListNode *cursor = head;
    int i = 0;
    while(cursor != NULL) {
        nodes[i] = cursor;
        i++; cursor = cursor->next;
    }
    nodes[i] = NULL;

    int target = i - n;
    if(target == 0) return head->next;

    nodes[target - 1]->next = nodes[target + 1];

    return head;
}

int main() {
    struct ListNode *head = (struct ListNode *)malloc(sizeof(struct ListNode));
    head->val = 1;
    head->next = NULL;

    struct ListNode *cursor = head;
    for(int i = 2; i <= 5; i++) {
        struct ListNode *node = (struct ListNode *)malloc(sizeof(struct ListNode));
        node->val = i;
        node->next = NULL;
        cursor->next = node;
        cursor = node;
    }

    struct ListNode *result = removeNthFromEnd(head, 5);
    while(result != NULL) {
        printf("%d ", result->val);
        result = result->next;
    }
    printf("\n");

    return 0;
}